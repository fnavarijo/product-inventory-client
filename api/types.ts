export * from '@/api/auth/types';
export * from '@/api/home/types';
export * from '@/api/products/types';
