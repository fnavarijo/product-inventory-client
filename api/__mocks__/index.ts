export const Products = {
  getAll: jest.fn(() => [{ id: 1 }, { id: 2 }, { id: 3 }]),
};
