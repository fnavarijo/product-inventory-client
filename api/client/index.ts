import { NuxtAxiosInstance } from '@nuxtjs/axios';

let axiosInstance: NuxtAxiosInstance;

export const initAxios = ({ axios }: { axios: NuxtAxiosInstance }) => {
  console.info('Axios instance set up');
  // @ts-ignore
  axiosInstance = axios;
};

export const getClient = () => ({
  axios: axiosInstance,
});
