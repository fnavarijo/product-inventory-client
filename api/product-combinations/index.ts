export * from '@/api/product-combinations/getAll';
export * from '@/api/product-combinations/getById';
