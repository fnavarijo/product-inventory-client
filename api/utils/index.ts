export * from '@/api/utils/resetProps';
export * from '@/api/utils/transformProduct';
export * from '@/api/utils/transformBy';
export * from '@/api/utils/renameProp';
