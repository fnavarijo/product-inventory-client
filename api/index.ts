import { initAxios } from './client';

export * as Auth from './auth';
export * as Home from './home';
export * as Products from './products';
export * as ProductCombinations from './product-combinations';

export {
  initAxios,
};
