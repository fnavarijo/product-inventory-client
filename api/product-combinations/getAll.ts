import map from 'lodash/fp/map';
import get from 'lodash/fp/get';
import pick from 'lodash/fp/pick';
import flow from 'lodash/fp/flow';

import { getClient } from '@/api/client';

import { Dictionary } from '@/@types/global';
import { ProductCombination } from '@/api/types';

export async function getAll (): Promise<Array<ProductCombination>> {
  const { data } = await getClient()
    .axios
    .get('/product-combinations');

  const pickFields = pick(['id', 'image']);
  const getImagesUrls = get('url');

  const transformImagesProperty = ({ image, ...rest }: { image: Array<Dictionary<any>> }) => ({ ...rest, image: getImagesUrls(image) });

  const transformToProductCombination = flow([
    pickFields,
    transformImagesProperty,
  ]);
  return map(transformToProductCombination)(data);
}
