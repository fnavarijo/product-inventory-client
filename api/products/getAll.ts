import map from 'lodash/fp/map';

import { getClient } from '@/api/client';
import { transformBaseProduct } from '@/api/utils';

import { Product } from '@/api/types';
import { Dictionary } from '@/@types/global';

export async function getAll (params?: Dictionary<any>): Promise<Array<Product>> {
  const { data } = await getClient()
    .axios
    .get('/products', { params });

  return map(transformBaseProduct)(data);
};
