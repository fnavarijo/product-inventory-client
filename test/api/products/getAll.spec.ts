import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { initAxios, Products } from '@/api';
import { mock, mockFilter } from '@/test/api/products/getAll.mock';
import { NuxtAxiosInstance } from '@nuxtjs/axios';

describe('Products.getAll function', () => {
  beforeAll(() => {
    const axiosInstance = new MockAdapter(axios);

    axiosInstance
      .onGet('/products', { params: { isFeatured: true } }).reply(200, mockFilter)
      .onGet('/products').reply(200, mock);

    initAxios({ axios: axios as NuxtAxiosInstance });
  });

  it('should fetch products list without filter', async () => {
    const products = await Products.getAll();

    expect(products[0]).toHaveProperty('id');
    expect(products[0]).toHaveProperty('name');
    expect(products[0]).toHaveProperty('category');
    expect(products[0]).toHaveProperty('featuredImage');
    expect(products[0]).toHaveProperty('isFeatured');

    expect(products).toMatchSnapshot();
  });

  it('should fetch products list with filter', async () => {
    const products = await Products.getAll({ isFeatured: true });

    expect(products[0]).toHaveProperty('id');
    expect(products[0]).toHaveProperty('name');
    expect(products[0]).toHaveProperty('category');
    expect(products[0]).toHaveProperty('featuredImage');
    expect(products[0]).toHaveProperty('isFeatured');

    expect(products).toMatchSnapshot();
  });

  it.todo('Test for checking images gallery');
});
