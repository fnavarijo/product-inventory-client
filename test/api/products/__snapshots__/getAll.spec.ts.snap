// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products.getAll function should fetch products list with filter 1`] = `
Array [
  Object {
    "category": "Zapatos - Casuales",
    "featuredImage": "https://res.cloudinary.com/dkvtpo8w1/image/upload/v1612318192/formal_shoes_8d4a69315b.png",
    "id": "601a06288d3ce427a235e38c",
    "isFeatured": true,
    "name": "Kyler OX",
  },
]
`;

exports[`Products.getAll function should fetch products list without filter 1`] = `
Array [
  Object {
    "category": "Zapatos - Casuales",
    "featuredImage": "https://res.cloudinary.com/dkvtpo8w1/image/upload/v1612318192/formal_shoes_8d4a69315b.png",
    "id": "601a06288d3ce427a235e38c",
    "isFeatured": false,
    "name": "Kyler OX",
  },
  Object {
    "category": "Ropa - Camisas",
    "featuredImage": "https://res.cloudinary.com/dkvtpo8w1/image/upload/v1612318199/shirt_75eb3b9fc2.png",
    "id": "601a072f39aaa829dd99de39",
    "isFeatured": false,
    "name": "Deportiva para ejercicio",
  },
]
`;
